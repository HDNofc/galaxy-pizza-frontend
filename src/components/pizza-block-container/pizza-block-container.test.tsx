import userEvent from '@testing-library/user-event';
import renderWithStore from 'test-utils/render-with-store';
import PizzaBlockContainer from './pizza-block-container';
import pizza from './__mocks__/index.js';

describe('<PizzaBlockContainer>', () => {
  it('отображается корректно', () => {
    const { container, getAllByRole, queryByRole, getByText, getByLabelText } = renderWithStore(
      <PizzaBlockContainer pizzaItem={pizza} />
    );

    expect(getByText(/Большая бонанза/i)).toBeInTheDocument();
    expect(getByText(/Оригинальная пицца с говядиной/i)).toBeInTheDocument();

    expect(getAllByRole('img', { hidden: true })[0]).toHaveAttribute(
      'src',
      'test-fat-crop-image.jpg'
    );
    expect(container.querySelector('source')).toHaveAttribute('srcset', 'test-fat-crop-image.webp');

    expect(getByText(/23/i)).toBeInTheDocument();
    expect(getByLabelText('23')).toBeChecked();
    expect(getByText(/30/i)).toBeInTheDocument();
    expect(getByText(/35/i)).toBeInTheDocument();
    expect(getByText(/40/i)).toBeInTheDocument();

    expect(getByText(/459 ₽/i)).toBeInTheDocument();
    expect(queryByRole('textbox', { hidden: true })).not.toBeInTheDocument();

    expect(container.firstChild).toMatchSnapshot();
  });

  it('переключение типа теста происходит корректно', () => {
    const { getByLabelText } = renderWithStore(<PizzaBlockContainer pizzaItem={pizza} />);

    expect(getByLabelText(/традиционное/i)).toBeChecked();
    expect(getByLabelText(/тонкое/i)).not.toBeChecked();

    userEvent.click(getByLabelText(/тонкое/i));

    expect(getByLabelText(/традиционное/i)).not.toBeChecked();
    expect(getByLabelText(/тонкое/i)).toBeChecked();
  });

  it('изменение изображения при переключении типа теста происходит корректно', () => {
    const { container, getByLabelText, getByRole } = renderWithStore(
      <PizzaBlockContainer pizzaItem={pizza} />
    );

    expect(getByRole('img', { hidden: true })).toHaveAttribute('src', 'test-fat-crop-image.jpg');
    expect(container.querySelector('source')).toHaveAttribute('srcset', 'test-fat-crop-image.webp');

    userEvent.click(getByLabelText(/тонкое/i));

    expect(getByRole('img', { hidden: true })).toHaveAttribute('src', 'test-slim-crop-image.jpg');
    expect(container.querySelector('source')).toHaveAttribute(
      'srcset',
      'test-slim-crop-image.webp'
    );
  });

  it('при переключении типа теста размер пиццы остается прежним', () => {
    const { getByLabelText, getByText, queryByText } = renderWithStore(
      <PizzaBlockContainer pizzaItem={pizza} />
    );

    // переключаем с 23см на 30см, т.к. 23см пиццы на тонком тесте нету в исходных данных
    userEvent.click(getByLabelText('30'));

    expect(getByText(/23/i)).toBeInTheDocument();
    expect(getByText(/30/i)).toBeInTheDocument();
    expect(getByLabelText('30')).toBeChecked();
    expect(getByText(/35/i)).toBeInTheDocument();
    expect(getByText(/40/i)).toBeInTheDocument();

    userEvent.click(getByLabelText(/тонкое/i));

    expect(queryByText(/23/i)).not.toBeInTheDocument();
    expect(getByText(/30/i)).toBeInTheDocument();
    expect(getByLabelText('30')).toBeChecked();
    expect(getByText(/35/i)).toBeInTheDocument();
    expect(getByText(/40/i)).toBeInTheDocument();
  });

  it('при переключении типа теста выбирается минимальный размер пиццы, если прежний размер пиццы недоступен', () => {
    const { getByLabelText, getByText, queryByText } = renderWithStore(
      <PizzaBlockContainer pizzaItem={pizza} />
    );

    expect(getByText(/23/i)).toBeInTheDocument();
    expect(getByLabelText('23')).toBeChecked();
    expect(getByText(/30/i)).toBeInTheDocument();
    expect(getByText(/35/i)).toBeInTheDocument();
    expect(getByText(/40/i)).toBeInTheDocument();

    userEvent.click(getByLabelText(/тонкое/i));

    expect(queryByText(/23/i)).not.toBeInTheDocument();
    expect(getByText(/30/i)).toBeInTheDocument();
    expect(getByLabelText('30')).toBeChecked();
    expect(getByText(/35/i)).toBeInTheDocument();
    expect(getByText(/40/i)).toBeInTheDocument();
  });

  it('при переключении типа теста изменяется цена', () => {
    const { getByLabelText, getByText } = renderWithStore(
      <PizzaBlockContainer pizzaItem={pizza} />
    );

    expect(getByLabelText(/традиционное/i)).toBeChecked();
    expect(getByLabelText('23')).toBeChecked();
    expect(getByText(/459 ₽/i)).toBeInTheDocument();

    userEvent.click(getByLabelText(/тонкое/i));
    expect(getByText(/669 ₽/i)).toBeInTheDocument();
  });

  it('при переключении размера изменяется цена', () => {
    const { getByLabelText, getByText } = renderWithStore(
      <PizzaBlockContainer pizzaItem={pizza} />
    );

    expect(getByLabelText(/традиционное/i)).toBeChecked();
    expect(getByLabelText('23')).toBeChecked();
    expect(getByText(/459 ₽/i)).toBeInTheDocument();

    userEvent.click(getByLabelText(/35/i));
    expect(getByText(/809 ₽/i)).toBeInTheDocument();
  });

  it('селект выбора кастомного бортика отображается в традиционном тесте на размере 30см+', () => {
    const { getByLabelText, queryByText, getByText } = renderWithStore(
      <PizzaBlockContainer pizzaItem={pizza} />
    );

    expect(getByLabelText(/традиционное/i)).toBeChecked();
    expect(getByLabelText('23')).toBeChecked();
    expect(queryByText(/Стандартный борт/i)).not.toBeInTheDocument();

    userEvent.click(getByLabelText(/30/i));
    expect(getByText(/Стандартный борт/i)).toBeInTheDocument();
  });

  it('селект выбора кастомного бортика не отображается в тонком тесте', () => {
    const { getByLabelText, queryByText } = renderWithStore(
      <PizzaBlockContainer pizzaItem={pizza} />
    );

    expect(getByLabelText(/традиционное/i)).toBeChecked();
    expect(getByLabelText('23')).toBeChecked();
    expect(queryByText(/Стандартный борт/i)).not.toBeInTheDocument();

    userEvent.click(getByLabelText(/30/i));
    userEvent.click(getByLabelText(/тонкое/i));
    expect(queryByText(/Стандартный борт/i)).not.toBeInTheDocument();
  });

  it('селект выбора кастомного бортика переключается корректно', () => {
    const { getByLabelText, queryByText, getByRole, getByText } = renderWithStore(
      <PizzaBlockContainer pizzaItem={pizza} />
    );

    userEvent.click(getByLabelText(/30/i));
    expect(queryByText(/стандартный борт/i)).toBeInTheDocument();

    const select = getByRole('textbox', { hidden: true });
    userEvent.click(select);

    userEvent.click(getByText(/сырный борт/i));
    expect(queryByText(/сырный борт/i)).toBeInTheDocument();
    expect(queryByText(/стандартный борт/i)).not.toBeInTheDocument();
  });

  it('при смене кастомного бортика изменяется цена', () => {
    const { getByLabelText, queryByText, getByRole, getByText } = renderWithStore(
      <PizzaBlockContainer pizzaItem={pizza} />
    );

    userEvent.click(getByLabelText(/30/i));
    expect(queryByText(/стандартный борт/i)).toBeInTheDocument();
    expect(getByText(/669 ₽/i)).toBeInTheDocument();

    const select = getByRole('textbox', { hidden: true });
    userEvent.click(select);

    userEvent.click(getByText(/сырный борт/i));
    expect(getByText(/868 ₽/i)).toBeInTheDocument();

    userEvent.click(select);
    userEvent.click(getByText(/колбасный борт/i));
    expect(getByText(/888 ₽/i)).toBeInTheDocument();
  });
});
