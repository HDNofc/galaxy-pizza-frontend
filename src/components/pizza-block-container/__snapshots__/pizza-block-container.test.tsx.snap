// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PizzaBlockContainer> отображается корректно 1`] = `
<article
  class="PizzaBlock-sc-1lztd8-0 fFmVty"
>
  <div
    class="ImagePlaceholder-sc-1lztd8-5 jnAVQG"
  >
    <div
      class="ImagePlaceholder-sc-1lq6aod-0"
    >
      <div
        class="Placeholder-sc-1lq6aod-1 eAAnAM"
      >
        <svg
          class="pizza-placeholder-image"
        >
          pizza-placeholder.svg
        </svg>
      </div>
      <div
        class="Content-sc-1lq6aod-2 iLQvsD"
      >
        <picture
          class="Picture-sc-1lztd8-3 dUIhhY"
        >
          <source
            srcset="test-fat-crop-image.webp"
            type="image/webp"
          />
          <img
            alt=""
            class="Image-sc-1lztd8-4 dvOTyR"
            src="test-fat-crop-image.jpg"
          />
        </picture>
      </div>
    </div>
  </div>
  <div
    class="PizzaInfo-sc-1lztd8-7 kExtID"
  >
    <div
      class="Category-sc-1lztd8-10 ZbJpL"
    >
      <svg>
        hot.svg
      </svg>
    </div>
    <h2
      class="Title-sc-1lztd8-1 iIWbhI"
    >
      Большая бонанза
    </h2>
    <p
      class="Description-sc-1lztd8-2 daMphU"
    >
      Оригинальная пицца с говядиной, ароматной ветчиной, двойной порцией хрустящего бекона, соусом "Барбекю", Моцареллой, мягким молодым сыром, томатами, жгучим перцем "Халапеньо"и луком
    </p>
    <div
      class="RadioGroup-sc-4trp2q-0 RePgg"
    >
      <div
        class="RadioGroupBox-sc-4trp2q-1 ffllcf"
        style="width: 50%; left: 0%;"
      />
      <div
        class="RadioGroupField-sc-18y3buo-0 dsLCSU"
      >
        <input
          checked=""
          class="Input-sc-18y3buo-2 hKbzCX"
          id="dough-14"
          name="dough-test-id-1"
          type="radio"
          value="fat"
        />
        <label
          class="Label-sc-18y3buo-1 loYKRY"
          for="dough-14"
        >
          Традиционное
        </label>
      </div>
      <div
        class="RadioGroupField-sc-18y3buo-0 dsLCSU"
      >
        <input
          class="Input-sc-18y3buo-2 hKbzCX"
          id="dough-15"
          name="dough-test-id-1"
          type="radio"
          value="slim"
        />
        <label
          class="Label-sc-18y3buo-1 ceQMEQ"
          for="dough-15"
        >
          Тонкое
        </label>
      </div>
    </div>
    <div
      class="Type-sc-1lztd8-8 bDHTSv"
    >
      <div
        class="RadioGroup-sc-4trp2q-0 RePgg"
      >
        <div
          class="RadioGroupBox-sc-4trp2q-1 ffllcf"
          style="width: 25%; left: 0%;"
        />
        <div
          class="RadioGroupField-sc-18y3buo-0 dsLCSU"
        >
          <input
            checked=""
            class="Input-sc-18y3buo-2 hKbzCX"
            id="size-1"
            name="size-test-id-1"
            type="radio"
            value="23"
          />
          <label
            class="Label-sc-18y3buo-1 loYKRY"
            for="size-1"
          >
            23
          </label>
        </div>
        <div
          class="RadioGroupField-sc-18y3buo-0 dsLCSU"
        >
          <input
            class="Input-sc-18y3buo-2 hKbzCX"
            id="size-2"
            name="size-test-id-1"
            type="radio"
            value="30"
          />
          <label
            class="Label-sc-18y3buo-1 ceQMEQ"
            for="size-2"
          >
            30
          </label>
        </div>
        <div
          class="RadioGroupField-sc-18y3buo-0 dsLCSU"
        >
          <input
            class="Input-sc-18y3buo-2 hKbzCX"
            id="size-3"
            name="size-test-id-1"
            type="radio"
            value="35"
          />
          <label
            class="Label-sc-18y3buo-1 ceQMEQ"
            for="size-3"
          >
            35
          </label>
        </div>
        <div
          class="RadioGroupField-sc-18y3buo-0 dsLCSU"
        >
          <input
            class="Input-sc-18y3buo-2 hKbzCX"
            id="size-4"
            name="size-test-id-1"
            type="radio"
            value="40"
          />
          <label
            class="Label-sc-18y3buo-1 ceQMEQ"
            for="size-4"
          >
            40
          </label>
        </div>
      </div>
    </div>
    <div
      class="ProductPurchase-sc-1lztd8-9 czSRRl"
    >
      <button
        class="Button-sc-1fut0te-0 kDjXIb"
        type="button"
      >
        <span
          class="Text-sc-1fut0te-1"
        >
          В корзину
        </span>
      </button>
      <span
        class="Price-sc-1lztd8-6 fgXguJ"
      >
        459 ₽
      </span>
    </div>
  </div>
</article>
`;
